mcproto/types/vec3.py

```diff
@@ -0,0 +1,187 @@
+        """Validate that the quaternion components are finite."""
+        if not math.isfinite(value):
+            raise ValueError(f"Quaternion components must be finite, got {value!r}")
+
+    x: float = field(validator=[validators.instance_of((float, int)), finite_validator.__get__(object)])
+    y: float = field(validator=[validators.instance_of((float, int)), finite_validator.__get__(object)])
+    z: float = field(validator=[validators.instance_of((float, int)), finite_validator.__get__(object)])
+
+    @override
+    def serialize_to(self, buf: Buffer) -> None:
+        buf.write_value(StructFormat.FLOAT, self.x)
+        buf.write_value(StructFormat.FLOAT, self.y)
+        buf.write_value(StructFormat.FLOAT, self.z)
+
+    @override
+    @classmethod
+    def deserialize(cls, buf: Buffer) -> Vec3:
+        x = buf.read_value(StructFormat.FLOAT)
+        y = buf.read_value(StructFormat.FLOAT)
+        z = buf.read_value(StructFormat.FLOAT)
+        return cls(x=x, y=y, z=z)
+
+    def serialize_to_double(self, buf: Buffer) -> None:
+        """Serialize the vector to a buffer using double precision.
+
+        .. seealso:: :meth:`serialize_to`
+        """
+        buf.write_value(StructFormat.DOUBLE, self.x)
+        buf.write_value(StructFormat.DOUBLE, self.y)
+        buf.write_value(StructFormat.DOUBLE, self.z)
+
+    @classmethod
+    def deserialize_double(cls, buf: Buffer) -> Vec3:
+        """Deserialize a vector from a buffer using double precision.
+
+        .. seealso:: :meth:`deserialize`
+        """
+        x = buf.read_value(StructFormat.DOUBLE)
+        y = buf.read_value(StructFormat.DOUBLE)
+        z = buf.read_value(StructFormat.DOUBLE)
+        return cls(x=x, y=y, z=z)
+
+    def __add__(self, other: Vec3) -> Vec3:
+        # Use the type of self to return a Vec3 or a subclass.
+        return type(self)(x=self.x + other.x, y=self.y + other.y, z=self.z + other.z)
+
+    def __sub__(self, other: Vec3) -> Vec3:
+        return type(self)(x=self.x - other.x, y=self.y - other.y, z=self.z - other.z)
+
+    def __neg__(self) -> Vec3:
+        return type(self)(x=-self.x, y=-self.y, z=-self.z)
+
+    def __mul__(self, other: float) -> Vec3:
+        return type(self)(x=self.x * other, y=self.y * other, z=self.z * other)
+
+    def __truediv__(self, other: float) -> Vec3:
+        return type(self)(x=self.x / other, y=self.y / other, z=self.z / other)
+
+    def to_position(self) -> Position:
+        """Convert the vector to a position."""
+        return Position(x=int(self.x), y=int(self.y), z=int(self.z))
+
+    def to_tuple(self) -> tuple[float, float, float]:
+        """Convert the vector to a tuple."""
+        return (self.x, self.y, self.z)
+
+    def to_vec3(self) -> Vec3:
+        """Convert the vector to a Vec3.
+
+        This function creates a new Vec3 object with the same components.
+        """
+        return Vec3(x=self.x, y=self.y, z=self.z)
```

Comment:
Add some note to the docstring here that this function is primarily intended to be used by subclasses of `Vec3` / `Position` class. It just looks really weird to see `Vec3.to_vec3` function.